Only set path in x64 mac installer when installing on x64

ericstj opened the pull request:
In other words don't set the path for x64 installer on ARM64

I find it odd that both the host and hostfxfr write to path today.  We could fix that if folks can say definitively if it is redundant.  For now I unified the duplicate scripts (one was missing a bugifx to ensure `paths.d` existed).

If we decide to have these scripts write to `/etc/dotnet/install_location` we might want to change how we calculate "installer arch" and do that by applying a template at build time.  I'm waiting to hear if we need to set that.

ericstj committed:
Only set path in x64 mac installer when installing on x64

In other words don't set the path for x64 installer on ARM64

ericstj commented:
@richlander let me know that he'd like us to write `/etc/dotnet/install_location`.  So I think I'll do that here.  As a result I'll refactor this PR to do both the path work and the writing to install_location.

ericstj commented:
@jkoritzinsky or @vitek-karas do either of you know why we would need the postinstall script in both host and hostfxr?  Do those ever get deployed seperately?

ericstj committed:
Remove postinstall script from hostfxr

There should be no need for both the host and hostfxr to set the path to dotnet.

Since the host installs dotnet, it should be the only package responsible for this.

am11 commented on src/installer/pkg/sfx/installers/osx_scripts/host-x64/postinstall:
```diff
@@ -0,0 +1,22 @@
+# A temporary fix for the permissions issue(s)
+chmod 755 $INSTALL_DESTINATION/dotnet
+
+if ["$_machine" == "amd64"] || ["$_machine" == "x86_x64"]; then
```
Given the POSIX shell shebang on line 1 `#!/bin/sh`, `==` (bash/zsh syntax) is not available in strict mode. Also missing spaces is a syntax error in all three shells.
```suggestion
if [ "$_machine" = "amd64" ] || [ "$_machine" = "x86_x64" ]; then
```

ericstj replied:
Thanks, actually refactored this a bit.  I'm a bash noob so I'm learning quite a bit.  Still testing, but could use another review shortly.

am11 replied:
In PoweShell, `[` is part of the language syntax where as in Unix shells, it's a built-in command with same functionality as `test`. Without the whitespace after `[`, it is iterpreted as `if test"$_machine = "amd64"`.

```sh
# [ is available in POSIX shell and its derivates (bash, ash, dash, zsh etc.)
bash-3.2$ command -v [
[

bash-3.2$ type -a [
[ is a shell builtin
[ is /bin/[

# ---

# [[ is only available in bash, ash, dash, zsh etc.
bash-3.2$ command -v [[
[[

bash-3.2$ type -a [[
[[ is a shell keyword
```

ericstj committed:
Make postinstall set install_location on mac

Also refactor script to use a template so that we don't need to fork the script.

ericstj committed:
fix some syntax errors in script

am11 commented on src/installer/pkg/sfx/installers/dotnet-host.proj:
```diff
@@ -78,4 +78,35 @@
+
+    <PropertyGroup>
+      <_UnameMachineRegex>$(InstallerTargetArchitecture)</_UnameMachineRegex>
+      <_UnameMachineRegex Condition="'$(InstallerTargetArchitecture)' == 'x64'">x64|x86_x64</_UnameMachineRegex>
```
```suggestion
      <_UnameMachineRegex Condition="'$(InstallerTargetArchitecture)' == 'x64'">x64|x86_64</_UnameMachineRegex>
```

ericstj replied:
Thanks, I was looking for official docs on this but had a hard time finding them.  I only had a M1 to test on.  Do you happen to know where the docs are?  I'm wondering if we need to handle more values.

ericstj replied:
Like aarch64 for arm, for example.  I only see the M1 return `arm64` but I'd really like to see docs on that to be certain.

am11 replied:
I am using BigSur x64, tested with `uname -m` which returns: `x86_64`. I think there is no agreement on these values across various Unix like OS. The possible values are `x86_64` and `amd64`. Some related resources:

* https://github.com/dotnet/runtime/blob/3409f73c7893f863297e5eb5d17241dcf0796733/eng/native/init-os-and-arch.sh
* CMake uses heuristics based on different `uname` flags, e.g. https://github.com/Kitware/CMake/blob/20b2bf00479ff656d0f6b772ba5673d1cb7007c4/Modules/CMakeDetermineSystem.cmake#L60
* https://stackoverflow.com/a/45125525/863980

ericstj replied:
I see, we check a these values:
https://github.com/dotnet/runtime/blob/3409f73c7893f863297e5eb5d17241dcf0796733/eng/native/init-os-and-arch.sh#L36-L40

I might match those for consistency in the repo.

I see wikipedia catalogs the values:
https://en.wikipedia.org/wiki/Uname
Looks like macos never changed the value once they started, so that's promising (this script only needs to run on macos).

None of these resources are what I would consider primary sources but it's better than nothing.  Thanks for sharing!

ericstj committed:
Update src/installer/pkg/sfx/installers/dotnet-host.proj

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

ericstj committed:
Refine uname regular expressions

vitek-karas commented on src/installer/pkg/sfx/installers/osx_scripts/host/postinstall:
```diff
@@ -10,9 +10,21 @@ INSTALL_DESTINATION=$2
+    mkdir -p /etc/dotnet    
+fi
+
+echo ${InstallerTargetArchitecture}=$INSTALL_DESTINATION | tee -a /etc/dotnet/install_location
```
Side note: For downlevel installers (3.1) we will need to write this without the architecture prefix. The arch prefix is only recognized by 6+ host.

ericstj replied:
Would be nice to put these requirements in the design doc cc @richlander

ericstj replied:
We had a chat about this and will further spec this behavior out.  Will write it up and then have a follow up change if the behavior needs to change here.

ericstj committed:
Merge branch 'main' into macPathX64

ericstj commented:
Since this commit was significantly changed since first approval, can I get a re-review?  We'd like to get this in and ported back to 6.0.
